src/wfs/src/main/java/org/geoserver/wfs/FeatureSizeFeatureCollection.java

```diff
@@ -0,0 +1,158 @@
+     * In some contexts, WFS-GetFeature requests need to precalculate the size 
+     * of the results, and (e.g. shapefile stores) the query is executed at 
+     * least twice causing two readings of the data source.
+     * </p>
+     */
+    static int FEATURE_CACHE_LIMIT = 
+        Integer.valueOf(System.getProperty("org.geoserver.wfs.getfeature.cachelimit", "16"));
+    
+    /**
+     * Allows to programmatically set the maximum number of cacheable features.
+     */
+    public static void setFeatureCacheLimit(int featureCacheLimit) {
+        FEATURE_CACHE_LIMIT = featureCacheLimit;
+    }
+
+    public FeatureSizeFeatureCollection(SimpleFeatureCollection delegate, FeatureSource<? extends FeatureType, ? extends Feature> source, Query query) {
+        super(delegate);
+        this.featureSource = source;
+        this.query = query;
+    }
+
+    class CachedWrappingFeatureIterator implements SimpleFeatureIterator {
+        
+        private List<SimpleFeature> featureCache;
+        private int featureIndex = 0;
+        
+        public CachedWrappingFeatureIterator(List<SimpleFeature> featureCache) {
+            this.featureCache = featureCache;
+        }
+        
+        @Override
+        public boolean hasNext() {
+            return featureIndex < featureCache.size();
+        }
+        
+        @Override
+        public SimpleFeature next() {
+            return featureCache.get(featureIndex++);
+        }
+        
+        @Override
+        public void close() {
+            featureIndex = 0;
+        }
+    }
+    
+    @Override
+    public SimpleFeatureIterator features() {
+        if (featureCache != null) {
+            return new CachedWrappingFeatureIterator( featureCache );
+        }
+        return super.features();
+    }
+    
+    @Override
+    public int size() {
+        if (featureCache != null) {
+            return featureCache.size();
+        }
+        if (FEATURE_CACHE_LIMIT > 0) {
+            FeatureIterator<? extends Feature> it = null;
+            
+            try {
+                int count = featureSource.getCount(query);
+                
+                if (count == 0) {
+                    featureCache = new ArrayList<SimpleFeature>();
+                    return count;
+                }
+                if (count > 0) {
+                    return count;
+                }
+                
+                // we have to iterate, save to cache to avoid later successive readings of data.
+                List<SimpleFeature> tempFeatureCache = new ArrayList<SimpleFeature>();
+                
+                // bean counting...
+                it = featureSource.getFeatures(query).features();
+                count = 0;
+                while (it.hasNext()) {
+                    SimpleFeature feature = (SimpleFeature) it.next();
+                    if (tempFeatureCache.size() < FEATURE_CACHE_LIMIT) tempFeatureCache.add(feature);
```

Comment:
Maybe I'm mis-reading it, but working this way, if there are more than limit features, don't we get exactly FEATURE_CACHE_LIMIT In the tempFeatureCache? Which would make the code below cache it?

Reply:
tempFeatureCache will be only used when the [count is lower than FEATURE_CACHE_LIMIT](https://github.com/geoserver/geoserver/pull/1355/files#diff-3855896b9c66cd8763ca54e143f382caR140), then we can only add features to this array upto this limit. Otherwise this collection wrapper has not effect because [featureCache is null](https://github.com/geoserver/geoserver/pull/1355/files#diff-3855896b9c66cd8763ca54e143f382caR104).